src/dividat-driver/firmware/main.go

```diff
@@ -86,29 +94,29 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
 			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
```

Comment:
Indeed as @guyonvarch suggested, trying to re-discover here with `_sensoControl._tcp` might not make sense. At this point we've already tried that in step one. I think it doesn't hurt, however unless a Senso could somehow switch back from DFU mode to normal mode between step 1 and 3 - which I don't think is the case - it seems unnecessary.

Reply:
Maybe something like this would make it more obvious?

```diff
diff --git a/src/dividat-driver/firmware/main.go b/src/dividat-driver/firmware/main.go
index aad158b..6b80743 100644
--- a/src/dividat-driver/firmware/main.go
+++ b/src/dividat-driver/firmware/main.go
@@ -62,7 +62,7 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 	} else {
 		// Discover controller address via mDNS
 		ctx, cancel := context.WithTimeout(parentCtx, 15*time.Second)
-		discoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
+		discoveredAddr, err := discoverSensoInNormalMode(deviceSerial, ctx)
 		cancel()
 		if err != nil {
 			fmt.Printf("Error: %s\n", err)
@@ -87,12 +87,12 @@ func Update(parentCtx context.Context, image io.Reader, deviceSerial *string, co
 		dfuHost = *configuredAddr
 	} else {
 		ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
-		discoveredAddr, err := discover("_sensoUpdate._udp", deviceSerial, ctx)
+		discoveredAddr, err := discoverSensoInDFUMode(deviceSerial, ctx)
 		cancel()
 		if err != nil {
 			// Try to discover boot controller via legacy identifier
 			ctx, cancel := context.WithTimeout(parentCtx, 60*time.Second)
-			legacyDiscoveredAddr, err := discover("_sensoControl._tcp", deviceSerial, ctx)
+			legacyDiscoveredAddr, err := discoverSensoInNormalMode(deviceSerial, ctx)
 			cancel()
 			if err == nil {
 				dfuHost = legacyDiscoveredAddr
@@ -172,6 +172,15 @@ func putTFTP(host string, port string, image io.Reader) error {
 	return nil
 }
 
+
+func discoverSensoInDFUMode(deviceSerial *string, ctx context.Context)(addr string, err error){
+  return discover("_sensoUpdate._udp", deviceSerial, ctx)
+}
+
+func discoverSensoInNormalMode(deviceSerial *string, ctx context.Context)(addr string, err error){
+  return discover("_sensoControl._tcp", deviceSerial, ctx)
+}
+
 func discover(service string, deviceSerial *string, ctx context.Context) (addr string, err error) {
 	resolver, err := zeroconf.NewResolver(nil)
 	if err != nil {
```

Reply:
Oh, I see. In firmware versions before 2.0.0.0 the bootloader advertised itself as `_sensoControl` instead of `_sensoUpdate`.

This is what is meant by "legacy" here, but we should add a more explicit comment.